library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java

```diff
@@ -51,291 +66,607 @@
+
+        FSWindowReplayWAL.FileDescriptor last = wal.fileDescriptors.get(sortedParts.last()).last();
+        if (last.isTmp) {
+          wal.tempPartFiles.put(last.part, last.filePath.toString());
+        }
+      }
     }
-    return storageAgent.load(operatorId, windowId);
   }
 
-  @Override
-  public void delete(int operatorId, long windowId) throws IOException
+  private void findFiles(FSWindowReplayWAL wal, int operatorId) throws IOException
   {
-    storageAgent.delete(operatorId, windowId);
+    String operatorDir = statePath + Path.SEPARATOR + operatorId;
+    Path operatorPath = new Path(operatorDir);
+    if (fileContext.util().exists(operatorPath)) {
+      RemoteIterator<FileStatus> walFilesIter = fileContext.listStatus(operatorPath);
+
+      while (walFilesIter.hasNext()) {
+        FileStatus fileStatus = walFilesIter.next();
+        FSWindowReplayWAL.FileDescriptor descriptor = FSWindowReplayWAL.FileDescriptor.create(fileStatus.getPath());
+        wal.fileDescriptors.put(descriptor.part, descriptor);
+      }
+    }
   }
 
-  @Override
-  public Map<Integer, Object> load(long windowId) throws IOException
+  private long findLargestRecoveryWindow(FSWindowReplayWAL wal, Long ceilingWindow) throws IOException
   {
-    Set<Integer> operators = replayState.get(windowId);
-    if (operators == null) {
-      return null;
-    }
-    Map<Integer, Object> data = Maps.newHashMap();
-    for (int operatorId : operators) {
-      data.put(operatorId, load(operatorId, windowId));
+    if (!wal.fileDescriptors.isEmpty()) {
+      FileSystemWAL.FileSystemWALReader reader = wal.getReader();
+
+      //to find the largest window, we only need to look at the last file.
+      NavigableSet<Integer> descendingParts = new TreeSet<>(wal.fileDescriptors.keySet()).descendingSet();
+      for (int part : descendingParts) {
+        FSWindowReplayWAL.FileDescriptor last = wal.fileDescriptors.get(part).last();
+        reader.seek(new FileSystemWAL.FileSystemWALPointer(last.part, 0));
+
+        long endOffset = -1;
+
+        long lastWindow = Stateless.WINDOW_ID;
+        Slice slice = readNext(reader);
+
+        while (slice != null) {
+          boolean skipComplete = skipNext(reader);
+          if (!skipComplete) {
+            //artifact not saved so this window was not finished.
+            break;
+          }
+          Slice windowSlice = slice;
+          long offset = reader.getCurrentPointer().getOffset();
+          slice = readNext(reader);  //either null, deleted or next window
+
+          if (slice == null || !slice.equals(DELETED)) {
+            //delete entry not found so window was not deleted
+            long window = Longs.fromByteArray(windowSlice.toByteArray());
+
+            if (ceilingWindow != null && window > ceilingWindow) {
+              break;
+            }
+            endOffset = offset;
+            lastWindow = window;
```

Comment:
@ilooner 
delete is handled by the else block from 283-284.
lastWindow and  endOffset are not at all modified for a window that is marked deleted.

I fail to see the problem. Can you please have another look?